src/dpq2/conv/time.d

```diff
@@ -13,11 +13,29 @@ import dpq2.oids;
 import dpq2.conv.to_d_types: throwTypeComplaint;
 import dpq2.exception;
 
-import std.datetime;
-import std.bitmanip: bigEndianToNative;
+import core.time;
+import std.datetime.date : Date, DateTime, TimeOfDay;
+import std.datetime.systime : LocalTime, SysTime, TimeZone, UTC;
+import std.bitmanip: bigEndianToNative, nativeToBigEndian;
 import std.math;
 import core.stdc.time: time_t;
 
+/// Returns value timestamp with time zone as SysTime
+SysTime binaryValueAs(T)(in Value v) @trusted
```

Comment:
But this method is to get SysTime from `TimeStampWithTZ` it just parses the time same way as `TimeStampWithoutTZ` and set is as UTC as it is in UTC.
It's not meant tu return SysTime from `OidType.TimeStampWithZone`which is checked there.

Reply:
I.e., you say that Postgres binary protocol supports `Timestamp with TZ` by transfer theese fields as `Timestamp without TZ` given to UTC?

Can you confirm this? (URL to documentation or code is preferred)

Reply:
This is probably the closest one (from https://www.postgresql.org/docs/9.6/static/datatype-datetime.html):

> For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's TimeZone parameter, and is converted to UTC using the offset for the timezone zone.

> When a timestamp with time zone value is output, it is always converted from UTC to the current timezone zone, and displayed as local time in that zone. To see the time in another time zone, either change timezone or use the AT TIME ZONE construct (see Section 9.9.3).

> Conversions between timestamp without time zone and timestamp with time zone normally assume that the timestamp without time zone value should be taken or given as timezone local time. A different time zone can be specified for the conversion using AT TIME ZONE.

So I assumed that in TEXT format, timestamp is returned with the local time TZ offset as defined in the DB.
But as the internal value is converted to and stored in UTC, I'm assuming that they don't bother for binary protocol and just send it as it is so in UTC.
This is also tested in integration tests.

Reply:
Wow! Thank! Isn't seen it.

So it is possible safetly create `toSysTime(immutable TimeZone tz = UTC())` method for `timestamptz` Values!

Reply:
But please don't add `PGtimestamptz` type alias - `SysTime` isn't equal to `timestamptz` because TZ data is omitted by server. TZ data also imporant in some applications as visual objective value for end users.

Reply:
Ok, I'll remove the alias. But internally it is the same regarding the TZ. They are not same in the precision and possible range of values.
SysTime internelly stores time in hnsecs in a long. PGtimestamptz stores usecs in long too.
Both differs in the date they count it, so ranges are different (SysTime has higher precision but because of it also 10 times lower min max range).
But both sores the ammount in UTC. When one works with SysTime and tries to show it to the user, it should be always converted to LocalTime - which will work either if SysTime from DB is in UTC or already in LocalTime.

Reply:
> Ok, I'll remove the alias. 

Stop! You persuaded me to leave this type :-)
This is just an alias and we can change it at any time for better types support. If we obtain type with better timestamptz support we can change it!

> it should be always converted to LocalTime 

This is done automatically?

Reply:
If we convert Value -> client SysTime -> Value, we will still have the same time in UTC. But PG doesn't store the information about TZ it was originally in - as it is not needed.
So when we are converting SysTime to Value, it is converted to UTC and then converted to binary format. And info of exact TZ is not there anymore, but can be easily converted to it back when needed.

> This is done automatically?

No this is not done currently. I can add the implicit conversion to LocalTime TZ, but it won't work for cases when original SysTime was in the other TZ. So we can't reconstruct the SysTime exactly as it was given to store in the DB, but important thing is, that still the time is correct and can be used for comparisons, etc. and easily converted back to desired TZ just by calling `toOtherTZ`.

Reply:
> But PG doesn't store the information about TZ it was originally in - as it is not needed.

PG stores info about TZ but isn't sends it by binary protocol. (you can check this by requesting `timestamptz` value trought text protocol)

Reply:
> SysTime internelly stores time in hnsecs in a long. PGtimestamptz stores usecs in long too.
Both differs in the date they count it, so ranges are different (SysTime has higher precision but because of it also 10 times lower min max range).
But both sores the ammount in UTC.

It will be good if user be warned about it. Or he will be able to choose to round values or throw an exception.

Reply:
> PG stores info about TZ but isn't sends it by binary protocol.

No it doesn't. It just stored it in UTC and returns it by default in your LocalTime TZ.

So for me for example `SELECT '2017-1-15 14:29:17.123456789+05'::timestamptz`
Returns `2017-01-15 10:29:17.123457+01`

This is from the PG docs:
> All timezone-aware dates and times are stored internally in UTC. They are converted to local time in the zone specified by the TimeZone configuration parameter before being displayed to the client.

Reply:
I have to admit my mistake! I apologize! :-(
I confused this with a special text(!) field in my own DB designed specifically for a storage timestamp with TZ offset in my project.
I apologize again!

It can be postulated that `timestamp` and `timestamptz` have equal size and internal storage format but timestamptz is already converted to UTC and `timestamp` have unknown TZ.

And `TimeStamp` can be `Typedef`ed from `TimeStampWithoutTZ` (or vice versa)!

Reply:
> If we convert Value -> client SysTime -> Value, we will still have the same time in UTC.

Sure! (But remember what this is true only for `timestamptz`.)